File: src/pinverify/verify.test.ts
import { describe, expect, it } from "vitest";
import { createServer } from "node:http";
import { check } from "./verify.js";
import { Client } from "../github/client.js";
import type { ActionUpdate } from "../diffparser/parser.js";

const sha40a = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
const sha40b = "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb";
const sha40c = "cccccccccccccccccccccccccccccccccccccccc";

function startServer(
  handler: (req: any, res: any) => void
): Promise<{ url: string; close: () => void }> {
  return new Promise((resolve) => {
    const server = createServer(handler);
    server.listen(0, "127.0.0.1", () => {
      const addr = server.address() as { port: number };
      resolve({
        url: `http://127.0.0.1:${addr.port}`,
        close: () => server.close(),
      });
    });
  });
}

describe("check", () => {
  it("no mismatch when SHA matches tag", async () => {
    const { url, close } = await startServer((req, res) => {
      if (req.url?.includes("/repos/actions/checkout/commits/v6.2.0")) {
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ sha: sha40a }));
        return;
      }
      res.writeHead(404);
      res.end();
    });

    try {
      const client = new Client("");
      client.setBaseURL(url);

      const updates: ActionUpdate[] = [
        {
          action: "actions/checkout",
          oldRef: "v6",
          newRef: sha40a,
          oldTag: "v6",
          newTag: "v6.2.0",
          file: ".github/workflows/ci.yml",
        },
      ];

      const mismatches = await check(client, updates);
      expect(mismatches).toHaveLength(0);
    } finally {
      close();
    }
  });

  it("detects mismatch", async () => {
    const { url, close } = await startServer((req, res) => {
      if (req.url?.includes("/repos/actions/checkout/commits/v6.2.0")) {
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ sha: sha40b }));
        return;
      }
      res.writeHead(404);
      res.end();
    });

    try {
      const client = new Client("");
      client.setBaseURL(url);

      const updates: ActionUpdate[] = [
        {
          action: "actions/checkout",
          oldRef: "v6",
          newRef: sha40a,
          oldTag: "v6",
          newTag: "v6.2.0",
          file: ".github/workflows/ci.yml",
        },
      ];

      const mismatches = await check(client, updates);
      expect(mismatches).toHaveLength(1);
      expect(mismatches[0].tag).toBe("v6.2.0");
      expect(mismatches[0].expectSHA).toBe(sha40b);
      expect(mismatches[0].update.newRef).toBe(sha40a);
    } finally {
      close();
    }
  });

  it("skips non-SHA ref", async () => {
    const { url, close } = await startServer((_req, res) => {
      res.writeHead(404);
      res.end();
    });

    try {
      const client = new Client("");
      client.setBaseURL(url);

      const updates: ActionUpdate[] = [
        {
          action: "actions/checkout",
          oldRef: "v4.0.0",
          newRef: "v4.1.0",
          oldTag: "v4.0.0",
          newTag: "v4.1.0",
          file: ".github/workflows/ci.yml",
        },
      ];

      const mismatches = await check(client, updates);
      expect(mismatches).toHaveLength(0);
    } finally {
      close();
    }
  });

  it("skips when no tag", async () => {
    const { url, close } = await startServer((_req, res) => {
      res.writeHead(404);
      res.end();
    });

    try {
      const client = new Client("");
      client.setBaseURL(url);

      const updates: ActionUpdate[] = [
        {
          action: "actions/checkout",
          oldRef: sha40a,
          newRef: sha40b,
          oldTag: "",
          newTag: "",
          file: ".github/workflows/ci.yml",
        },
      ];

      const mismatches = await check(client, updates);
      expect(mismatches).toHaveLength(0);
    } finally {
      close();
    }
  });

  it("deduplicates API calls", async () => {
    let calls = 0;
    const { url, close } = await startServer((req, res) => {
      if (req.url?.includes("/repos/actions/checkout/commits/v6.2.0")) {
        calls++;
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ sha: sha40a }));
        return;
      }
      res.writeHead(404);
      res.end();
    });

    try {
      const client = new Client("");
      client.setBaseURL(url);

      const updates: ActionUpdate[] = [
        { action: "actions/checkout", oldRef: "", newRef: sha40a, oldTag: "", newTag: "v6.2.0", file: ".github/workflows/ci.yml" },
        { action: "actions/checkout", oldRef: "", newRef: sha40a, oldTag: "", newTag: "v6.2.0", file: ".github/workflows/build.yml" },
        { action: "actions/checkout", oldRef: "", newRef: sha40a, oldTag: "", newTag: "v6.2.0", file: ".github/workflows/test.yml" },
      ];

      const mismatches = await check(client, updates);
      expect(mismatches).toHaveLength(0);
      expect(calls).toBe(1);
    } finally {
      close();
    }
  });

  it("handles multiple actions", async () => {
    const { url, close } = await startServer((req, res) => {
      if (req.url?.includes("/repos/actions/checkout/commits/v6.2.0")) {
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ sha: sha40a }));
        return;
      }
      if (req.url?.includes("/repos/actions/setup-go/commits/v5.1.0")) {
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ sha: sha40c }));
        return;
      }
      res.writeHead(404);
      res.end();
    });

    try {
      const client = new Client("");
      client.setBaseURL(url);

      const updates: ActionUpdate[] = [
        { action: "actions/checkout", oldRef: "", newRef: sha40a, oldTag: "", newTag: "v6.2.0", file: ".github/workflows/ci.yml" },
        { action: "actions/setup-go", oldRef: "", newRef: sha40b, oldTag: "", newTag: "v5.1.0", file: ".github/workflows/ci.yml" },
      ];

      const mismatches = await check(client, updates);
      expect(mismatches).toHaveLength(1);
      expect(mismatches[0].update.action).toBe("actions/setup-go");
    } finally {
      close();
    }
  });

  it("gracefully handles API errors", async () => {
    const { url, close } = await startServer((_req, res) => {
      res.writeHead(404, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ message: "Not Found" }));
    });

    try {
      const client = new Client("");
      client.setBaseURL(url);

      const updates: ActionUpdate[] = [
        { action: "actions/checkout", oldRef: "", newRef: sha40a, oldTag: "", newTag: "v99.0.0", file: ".github/workflows/ci.yml" },
      ];

      const mismatches = await check(client, updates);
      expect(mismatches).toHaveLength(0);
    } finally {
      close();
    }
  });
});

File: pkg/pinverify/verify_test.go
package pinverify

import (
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/vvoland/gha-pin-diff/pkg/diffparser"
	"github.com/vvoland/gha-pin-diff/pkg/github"
)

const (
	sha40a = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
	sha40b = "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
	sha40c = "cccccccccccccccccccccccccccccccccccccccc"
)

func TestCheckMatch(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/repos/actions/checkout/commits/v6.2.0", func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"sha":"` + sha40a + `"}`))
	})

	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := github.NewClient("")
	client.SetBaseURL(srv.URL)

	updates := []diffparser.ActionUpdate{
		{
			Action: "actions/checkout",
			OldRef: "v6",
			NewRef: sha40a,
			OldTag: "v6",
			NewTag: "v6.2.0",
			File:   ".github/workflows/ci.yml",
		},
	}

	mismatches := Check(t.Context(), client, updates)
	if len(mismatches) != 0 {
		t.Errorf("expected no mismatches, got %d: %+v", len(mismatches), mismatches)
	}
}

func TestCheckMismatch(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/repos/actions/checkout/commits/v6.2.0", func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"sha":"` + sha40b + `"}`))
	})

	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := github.NewClient("")
	client.SetBaseURL(srv.URL)

	updates := []diffparser.ActionUpdate{
		{
			Action: "actions/checkout",
			OldRef: "v6",
			NewRef: sha40a, // does not match sha40b
			OldTag: "v6",
			NewTag: "v6.2.0",
			File:   ".github/workflows/ci.yml",
		},
	}

	mismatches := Check(t.Context(), client, updates)
	if len(mismatches) != 1 {
		t.Fatalf("expected 1 mismatch, got %d", len(mismatches))
	}
	m := mismatches[0]
	if m.Tag != "v6.2.0" {
		t.Errorf("Tag = %q, want v6.2.0", m.Tag)
	}
	if m.ExpectSHA != sha40b {
		t.Errorf("ExpectSHA = %q, want %q", m.ExpectSHA, sha40b)
	}
	if m.Update.NewRef != sha40a {
		t.Errorf("Update.NewRef = %q, want %q", m.Update.NewRef, sha40a)
	}
}

func TestCheckSkipsNonSHARef(t *testing.T) {
	srv := httptest.NewServer(http.NotFoundHandler())
	defer srv.Close()

	client := github.NewClient("")
	client.SetBaseURL(srv.URL)

	updates := []diffparser.ActionUpdate{
		{
			Action: "actions/checkout",
			OldRef: "v4.0.0",
			NewRef: "v4.1.0", // not a SHA
			OldTag: "v4.0.0",
			NewTag: "v4.1.0",
			File:   ".github/workflows/ci.yml",
		},
	}

	mismatches := Check(t.Context(), client, updates)
	if len(mismatches) != 0 {
		t.Errorf("expected no mismatches for non-SHA ref, got %d", len(mismatches))
	}
}

func TestCheckSkipsNoTag(t *testing.T) {
	srv := httptest.NewServer(http.NotFoundHandler())
	defer srv.Close()

	client := github.NewClient("")
	client.SetBaseURL(srv.URL)

	updates := []diffparser.ActionUpdate{
		{
			Action: "actions/checkout",
			OldRef: sha40a,
			NewRef: sha40b,
			File:   ".github/workflows/ci.yml",
		},
	}

	mismatches := Check(t.Context(), client, updates)
	if len(mismatches) != 0 {
		t.Errorf("expected no mismatches when no tag, got %d", len(mismatches))
	}
}

func TestCheckDeduplicatesAPICalls(t *testing.T) {
	calls := 0
	mux := http.NewServeMux()
	mux.HandleFunc("/repos/actions/checkout/commits/v6.2.0", func(w http.ResponseWriter, _ *http.Request) {
		calls++
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"sha":"` + sha40a + `"}`))
	})

	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := github.NewClient("")
	client.SetBaseURL(srv.URL)

	// Same action+tag in multiple files.
	updates := []diffparser.ActionUpdate{
		{Action: "actions/checkout", NewRef: sha40a, NewTag: "v6.2.0", File: ".github/workflows/ci.yml"},
		{Action: "actions/checkout", NewRef: sha40a, NewTag: "v6.2.0", File: ".github/workflows/build.yml"},
		{Action: "actions/checkout", NewRef: sha40a, NewTag: "v6.2.0", File: ".github/workflows/test.yml"},
	}

	mismatches := Check(t.Context(), client, updates)
	if len(mismatches) != 0 {
		t.Errorf("expected no mismatches, got %d", len(mismatches))
	}
	if calls != 1 {
		t.Errorf("expected 1 API call (deduplicated), got %d", calls)
	}
}

func TestCheckMultipleActions(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/repos/actions/checkout/commits/v6.2.0", func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"sha":"` + sha40a + `"}`))
	})
	mux.HandleFunc("/repos/actions/setup-go/commits/v5.1.0", func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		// Mismatch: tag resolves to sha40c but pin says sha40b.
		_, _ = w.Write([]byte(`{"sha":"` + sha40c + `"}`))
	})

	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := github.NewClient("")
	client.SetBaseURL(srv.URL)

	updates := []diffparser.ActionUpdate{
		{Action: "actions/checkout", NewRef: sha40a, NewTag: "v6.2.0", File: ".github/workflows/ci.yml"},
		{Action: "actions/setup-go", NewRef: sha40b, NewTag: "v5.1.0", File: ".github/workflows/ci.yml"},
	}

	mismatches := Check(t.Context(), client, updates)
	if len(mismatches) != 1 {
		t.Fatalf("expected 1 mismatch, got %d", len(mismatches))
	}
	if mismatches[0].Update.Action != "actions/setup-go" {
		t.Errorf("mismatch action = %q, want actions/setup-go", mismatches[0].Update.Action)
	}
}

func TestCheckAPIError(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/", func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"message":"Not Found"}`))
	})

	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := github.NewClient("")
	client.SetBaseURL(srv.URL)

	updates := []diffparser.ActionUpdate{
		{Action: "actions/checkout", NewRef: sha40a, NewTag: "v99.0.0", File: ".github/workflows/ci.yml"},
	}

	// API errors should not produce mismatches (graceful degradation).
	mismatches := Check(t.Context(), client, updates)
	if len(mismatches) != 0 {
		t.Errorf("expected no mismatches on API error, got %d", len(mismatches))
	}
}
